feat: add deleted_for_me field in message response

daksh-r opened the pull request:
## Goal

handle delete-for-me scenario in messages:

1. hard delete and delete-for-me are mutually excluding
2. introduce options object as a second parameter to `client.deleteMessage()` method
3. enrich the delete message response message object with `type` `'deleted'` and `deleted_for_me` boolean flag

MartinCupela commented:
@daksh-r if I understand it well, we still need to update `client.deleteMessage()` and `client._deleteMessage()` implementation. This will also have impact on offline support (right @isekovanic?).

for the user that deleted it for himself: 

What will be the `message.type` value? Will it be `deleted` or `regular`?
What will be the `message.text` value?

MartinCupela commented:
@daksh-r  why the WS event has the following format:

```
type: "message.deleted"
deleted_for_me: true
message: { … }
```

but the message object will contain `deleted_for_me` too (as declared in `MessageResponseBase`)?

Isn't that duplication? Is it necessary to have `deleted_for_me` at the root of the WS event object?

daksh-r commented:
> @daksh-r why the WS event has the following format:
> 
> ```
> type: "message.deleted"
> deleted_for_me: true
> message: { … }
> ```
> 
> but the message object will contain `deleted_for_me` too (as declared in `MessageResponseBase`)?
> 
> Isn't that duplication? Is it necessary to have `deleted_for_me` at the root of the WS event object?

in the ws event, we don't send `deleted_for_me` inside the message {...} , it only gets sent once at the root of the `message.deleted` event, when a client subsequently tries to fetch the message(via the queryChannels or getMessage or getManyMessages endpoints) that's when we populate the `deleted_for_me` field in there.

MartinCupela commented on src/types.ts:
```diff
@@ -1455,6 +1456,7 @@ export type Event = CustomEventData & {
   connection_id?: string;
   // event creation timestamp, format Date ISO string
   created_at?: string;
+  deleted_for_me?: boolean;
```
Is it necessary to add this property to the WS event payload root if it is already contained inside the `message` object? I suppose that this would be message.deleted event that carries the `message`.

MartinCupela replied:
I see that I already pointed this out and see that the `message` inside the WS event is not getting the property.

MartinCupela commented on test/unit/client.test.js:
```diff
@@ -1019,6 +1058,171 @@ describe('message deletion', () => {
+		);
+	});
+
+	it('should return normal message when other user fetches the message', async () => {
```
It seems that this test just tests, whether the spy returns the value it was told to return.

MartinCupela commented on test/unit/client.test.js:
```diff
@@ -1019,6 +1058,171 @@ describe('message deletion', () => {
+		expect(deletingUserEventSpy).toHaveBeenCalledOnce();
+	});
+
+	it('should return deleted_for_me=true when deleting user fetches the message', async () => {
```
It seems that this test just tests, whether the spy returns the value it was told to return.

MartinCupela commented on test/unit/client.test.js:
```diff
@@ -1019,6 +1058,171 @@ describe('message deletion', () => {
+		);
+	});
+
+	it('should handle message.deleted event with deleted_for_me=true for deleting user', () => {
```
I believe this test is not testing what it declares. The test itself invokes WS event handlers for `deletingUserClient` but not for `otherUserClient`. It is normal that `otherUserEventSpy` will not be called if it is not registered with `deletingUserClient`. I think this test can be removed.
